\documentclass[a4paper]{article}
\usepackage{pdfpages}
\usepackage[latin9]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{graphicx}
\usepackage{color}
\usepackage{Sweave}
\usepackage{pdfcolmk}
\usepackage{multicol}
\usepackage{ifthen}
\usepackage{multido}
\usepackage[labelformat = empty]{caption}
\usepackage{fancyhdr}
\usepackage[top=1.2cm, bottom=0.5cm, right=1.3cm, left=1.3cm]{geometry}
\usepackage{lastpage}
\pagestyle{fancy}
\usepackage{fancyvrb}
\usepackage{etoolbox}
\usepackage{hyperref}
\hypersetup{
    colorlinks=true, %set true if you want colored links
    linktoc=all,     %set to all if you want both sections and subsections linked
    linkcolor=blue,  %choose some color if you want links to stand out
}
\makeatletter
\patchcmd{\FV@ListVSpace}{\@topsepadd\topsep}{}{}  %% Reduce space between Sin/output
\makeatother
\fancyhf{}
\rhead{\date\today \hspace{0.2cm}- \thepage/\pageref{LastPage}}
\renewcommand{\footrulewidth}{0pt}
\newcommand{\bslash}{\textbackslash}

\begin{document}

\setkeys{Gin}{width = 0.2\textwidth}  %% Auto position of the figures

\setlength\parindent{0pt}
\SweaveOpts{tidy = FALSE, keep.source = TRUE, keep.space = FALSE, keep.blank.space = FALSE, keep.comment = TRUE}
\DefineVerbatimEnvironment{Sinput}{Verbatim}{formatcom = {\color[rgb]{0, 0, 0.56}}}
\DefineVerbatimEnvironment{Soutput}{Verbatim}{formatcom = {\color[rgb]{0.56, 0, 0}}}

<< options, echo = FALSE >>=
options(prompt = ">", continue = "...", width = 100)
@

\title{Condensine implication in gene expression}
\author{Claire Burny, Cl√©mence Hocquet, Xavier Robellet, Laurent Modolo, Pascal Bernard}
\maketitle
\tableofcontents
\newpage

\section{Settings and sourcing useful functions}

When and where we are :
<<dir-and-date, echo = FALSE>>=
date();
R.Version()$version.string;
getwd();
require(formatR) # indentation with 2 blank spaces and lines of max 70 char
@

We source here useful functions and load necessary packages and tools. See the material and methods for more information on the data analysis protocol and the version fo the tools used.

\setkeys{Gin}{width = 0.8\textwidth}  

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Data pre-processing}

Data were sequenced on 1 lane on the same flowcell with an Illumina Hiseq platform 50-read length single-end in IGBMC (S17017 project). This was RNA seq experiments using a Ribozero kit depletion.\\
Each strain was sequenced in triplicates: 
\begin{itemize}
  \item WT: samples 1, 6, 11
  \item cut14-208 mutant: samples 2, 7, 12
  \item rrp6D mutant: samples 5, 10, 15
  \item cut14-208/cdc15-118 double mutant: samples 3, 8, 13
  \item cdc15-118 mutant: samples 4, 9, 14
\end{itemize}

Samples from 59,340,962 to 93,363,639 reads after adapters removal were provided. 
The encoding of quality values is Sanger / Illumina 1.9 format.\\
See doc directory for reports produced by IGBMC. 

% ----------------
\subsection{FastQC}

\begin{itemize}
  \item adapter trimming. This was performed by IGBMC and not done, indeed there is no contamination in adapter as regards FASTQ reports. 
  \item quality of sequencing. It is good quality data as regards mean(QPhred)>30 whathever the position within the read on both ends. There is potential high level of library duplication for some samples (presence of observations in the last duplicate bin plot).
  \item bp content. First 12 bp are biased in bp content probably due to experimental protocole (random hexamer fixation on RNA fragments), this does not required trimming. 3 samples (cut14 repl1, rrp6D repl3 and cdc15 repl1) display, as raw data and after trimming, 2 modes in GC\% distribution; this reflects rRNA sequences after some BLAST attempts. Indeed according to https://www.pombase.org/status/statistics, rRNA contains 48\% of GC compared to protein coding genes (36\%). This could also reflects tRNA content (54\% of GC). These 3 samples also display a high content of overrepresented sequences. Note that cdc15 repl2 also displays a high content of overrepresented sequences.  
\end{itemize}

\includegraphics{{../results/quality_control/fastqc_per_sequence_gc_content_plot}.png}\\
\includegraphics{{../results/quality_control/fastqc_overrepresented_sequencesi_plot}.png}


% ------------------------
\subsection{Data cleaning}

Laurent Modolo performed quality control on PSMN. We used cutadapt to trim tails with quality lower than 20. No other filter was applied. Trimmed reads are located in \textcolor{blue}{results/quality\_control/trimming} directory.

\begin{tabular}{c|c|c|c|c|c}
sample & initial & after adapter removal by IGBMC & \% bases over 30 & after cleaning & \% GC\\
\hline
1 - wt repl1 & 59,342,221 & 59,340,962 & 97.22 & 59,340,962 & 41\\
6 - wt repl2 & 73,560,652 & 73,556,476 & 97.28 & 73,556,476 & 41\\
11 - wt repl3 & 71,364,436 & 71,362,250 & 97.31 & 71,362,250 & 41\\
\hline
2 - cut14-208 repl1 & 71,036,268 & 71,035,324 & 97.36 & 71,035,324  & 43\\
7 - cut14-208 repl2 & 80,549,007 & 80,546,568 & 97.31 &  80,546,568& 41\\
12 - cut14-208 repl3 & 80,171,863 & 80,170,034 & 97.38 &  80,170,034 & 41\\
\hline
3 - cut14-208  cdc15-118 repl1 & 78,987,460 & 78,984,731 & 97.16 &78,984,731 & 41\\
8 - cut14-208  cdc15-118 repl2 & 72,204,300 & 72,200,808 & 97.26 & 72,200,808 & 41\\
13 - cut14-208  cdc15-118 repl3 & 84,285,814 & 84,240,516 & 97.06 & 84,285,814 & 41\\
\hline
4 - cdc15-118 repl1 & 77,862,467 & 77,859,813 & 97.26 &  77,859,813& 43\\
9 - cdc15-118 repl2 &71,883,098 & 71,880,239 & 97.3 &  71,880,239& 42\\
14 - cdc15-118 repl3 & 85,663,522 & 85,661,211 & 97.22 &  85,661,211  & 41\\
\hline
5 - rrp6D repl1 & 76,816,322 & 76,809,781 & 97.24 & 76,809,781  & 40\\
10 - rrp6D repl2 & 93,367,340 & 93,363,639 & 97.13 &  93,363,639 & 41\\
15 - rrp6D repl3 & 78,292,366 & 78,288,309 & 97.36 & 78,288,309 & 43\\
\end{tabular}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Mapping and quantification}

% ----------------------------------------
\subsection{Pick S. pombe reference genome}

S. pombe ASM294v2.30 fasta and gff3 were downloaded resp. with wget from \texttt{ftp://ftp.ebi.ac.uk/pub/databases/pombase/pombe/Chromosome\_Dumps/Schizosaccharomyces\_pombe.ASM294v2.30.dna.genome.fa.gz} and \texttt{ftp://ftp.ebi.ac.uk/pub/databases/pombase/pombe/Chromosome\_Dumps/gff3/schizosaccharomyces\_pombe.chr.gff3} on 9th August 2017.

I: 5,579,133bp\\
II: 4,539,804bp\\
III: 2,452,883bp


% ----------------------------------------------
\subsection{Mapping to S. pombe reference genome}

Mapping was performed by Laurent Modolo on PSMN from the trimmed FASTQ files; outputs of mapping are saved in \textcolor{blue}{results/mapping/mapped} and classified by strain (\texttt{2017\_10\_26} sorted BAM files).

\begin{tabular}{c|c|c|c|c}
sample & aligned concordantly 0 times & exactly 1 time & > 1 time & \% mapped\\
\hline
1 - wt repl1 & 931,061 (1.57\%) & 51,861,837 (87.4\%) & 6,548,064 (11.03\%) & 98.43\%\\
6 - wt repl2 &  1,404,250 (1.91\%) & 63,481,938 (86.30\%) & 8,670,288 (11.79\%) & 98.09\%\\
11 - wt repl3 & 1,050,119 (1.47\%) & 63,180,210 (88.53\%)& 7,131,921 (9.99\%) & 98.53\%\\
\hline
2 - cut14-208 repl1 & 718,161 (1.01\%)  & 40,284,357 (56.71\%) & \textcolor{red}{30,032,806 (42.28\%)}& 98.99\%\\
7 - cut14-208 repl2 & 1,210,643 (1.50\%) & 70,774,395 (87.87\%) & 8,561,530 (10.63\%) & 98.5\%\\
12 - cut14-208 repl3 & 1,073,503 (1.34\%) & 71,341,832 (88.99\%) & 7,754,699 (9.67\%)& 98.66\%\\
\hline
3 - cut14-208  cdc15-118 repl1 & 1,203,716 (1.52\%) & 67,301,903 (85.21\%) & 10,479,112 (13.27\%) & 98.48\%\\
8 - cut14-208  cdc15-118 repl2 & 1,084,581 (1.50\%)  &  63,804,070 (88.37\%)& 7,312,157 (10.13\%) & 98.5\%\\
13 - cut14-208  cdc15-118 repl3 &  1,248,135 (1.48\%)  & 73,929,409 (87.76\%) &  9,062,972 (10.76\%) & 98.52\%\\
\hline
4 - cdc15-118 repl1 & 848,597 (1.09\%) &  41,178,330 (52.89\%) &  \textcolor{red}{ 35,832,886 (46.02\%) )} & 98.91\%\\
9 - cdc15-118 repl2 & 1,171,116 (1.63\%) & 52,883,443 (73.57\%) & \textcolor{red}{17,825,680 (24.80\%)} & 98.37\%\\
14 - cdc15-118 repl3 & 1,125,054 (1.31\%) & 75,045,171 (87.61\%) & 9,490,986 (11.08\%) & 98.69\%\\
\hline
5 - rrp6D repl1 & 1,389,131 (1.81\%) & 66,121,823 (86.09\%)& 9,298,827 (12.11\%) & 98.19\%\\
10 - rrp6D repl2 &  1,612,648 (1.73\%) & 77,072,084 (82.55\%)  & 14,678,907 (15.72\%) & 98.27\%\\
15 - rrp6D repl3 & 926,017 (1.18\%)& \textcolor{red}{37,437,085 (47.82\%))} &  \textcolor{red}{39,925,207 (51.00\%)} & 98.82\%\\
\end{tabular}


%%%%%%%%%%%%%%%%%%%%%%%%%
\section{DESeq2 analysis}

In presence of condensine defects, 2 phenotypes are observed: 1) an overexpression of genes and 2) readthrough for some genes. 4 mutants are used to study direct and indirect effects of condensine defects (+ a WT strain):
\begin{itemize}
  \item cut14: condensine faulty
  \item cdc15: no cytokinesis 
  \item cdc15-cut14: double mutant, no daughter cell
  \item rrp6
\end{itemize}

Several analysis need to be done:
\begin{itemize}
  \item cut14-208 vs wt: phenotype 1 observed
  \item cdc15-118 vs wt: no more phenotype 1 observed: cdc15 suppresses effects of cut14.
  \item cut14-208\_cdc15-118 vs wt: no more phenotype 1 observed: cdc15 suppresses effects of cut14.
  \item cut14-208\_cdc15-118 vs cut14-208 / cut14-208\_cdc15-118 vs cdc15-118: should be equivalent if cdc15 suppresses effect of cut14.
  \item rrp6D vs wt
  \item rrp6D vs cut14-208
\end{itemize}

Each time, conditions are in triplicates. Annotation is saved in \texttt{/data/2017\_10\_03\_annotation.csv}.\\
Detailed results were saved in \texttt{results/DESeq2\_analysis/}. Within each results directory, a CSV file indicates outputs for all genes ordererd by increasing adjusted p-val value, so called \texttt{deseq2\_shrinked\_analysis\_cut14-208\_cdc15-118\_vs\_cut14-208.csv} for the analysis of the double mutant vs cut14 for instance.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Meeting 09.26.2017}

\begin{itemize}
  \item Validate on what we work for RNASeq\\
  \textcolor{green}{Annotations from PomBase for ncRNA\_gene, pseudogene, gene, snoRNA\_gene, snRNA\_gene on chromosomes I, II and III. The gff used is located in \texttt{data/2017\_10\_03\_selected\_features\_s\_pombe.gff}. The gff used for tRNAs is in \texttt{data/2017\_10\_03\_tRNAs\_s\_pombe.gff}.} Note that inputs for RSEM must be in gtf2 format.

  \item Do RNASeq analysis with adding missing replicates.\\
  \textcolor{green}{This was done on 2017.10.03.}
  
  \item Validate comparisons to be done with rrp6 mutant.\\
  \textcolor{green}{Venn diagrams will be done by Clemence Hocquet and Pascal Bernard. }
  
  \item Validate figures of density of up-regulated genes.\\
  \textcolor{green}{Density of up/down and equally expressed genes were plotted.}

  \item Validate distance between genes and tRNA + reformulate questions as regards the test wanted.\\
    \textcolor{green}{A KS unilateral test as well as a GLM quasi Poisson were done and pvalues were reported on plots.}

  \item Readthrough: now we have the expressiong per gene, need to compute gene by gene profiles with different sizes of windows. 
  
\end{itemize}



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Readthrough detection}

The protocol is described in details in the material and methods. 

% ---------------------------------------------
\subsection{Detection of raw readthrough events}

\begin{enumerate}
  \item \textcolor{blue}{src/prepare\_readthrough\_detection.sh}:
sorted BAM files obtained after bowtie2 mapping were saved in two sub files - forward and reverse BAM - to make 2 peaks calling (one per strand) per cut14 and rrp6D mutant. Intersed BAM on the corresponding features (+ if reverse BAM and - if forward BAM) have the extion \texttt{\_without}. Outputs are in \texttt{results/mapping/mapped/} and in the corresponding subdirectory. I also generated the mappability profile.
  \item \textcolor{blue}{src/preprocess\_MUSIC.sh} (PSMN):
the previous BAM files are preprocessed for sorting, removal of duplicates with MUSIC.
  \item \textcolor{blue}{src/peak\_calling\_MUSIC.sh} (PSMN):
  peak calling is performed with MUSIC by pooling all triplicates samples from the mutant of interest and compared to the pooled WT samples. Outputs from MUSIC are saved in \texttt{results/readthrough\_analysis/Output\_Music} within the corresponding directories. 
  \item \textcolor{blue}{src/filter\_readthrough\_peaks.R}:
 raw detected peaks are filtered after intersection with the corresponding stranded annotation (output for forward detection are intersected with - annotation considering again the strand-specific protocol of library preparation) using \textcolor{blue}{src/substract\_2\_bed.sh}. Only peaks which frontiers correspond to the 3' of the feature are kept. Two annotation are thus generated, based on rrp6 vs WT peaks detection and based on cut14 vs WT peaks detection:\\
annot\_readthrough\_with\_other\_genes\_rrp6D\_wt\_based\_cut14\_forward\_reverse.bed\ (readthrough and their parent, 1 line for the parent and the next line for readthrough + other genes)\
annot\_readthrough\_with\_other\_genes\_cut14\_wt\_based\_rrp6D\_forward_reverse.bed
annot\_readthrough\_rrp6D\_wt\_based\_cut14\_forward\_reverse.bed (readthrough and their parent only, 1 line for the parent and the next line for readthrough)\\
annot\_readthrough\_cut14\_wt\_based\_rrp6D\_forward\_reverse.bed
  \item \textcolor{blue}{kallisto\_quantif.sh}: quantification is made from annotation of all genes for classic DEA and from annotation of reathrough and their parents + other genes. The last quantification will allow to select putative readthrough with sleuth. Quantification are stored in \texttt{results/kallisto\_quantif/} with \texttt{\_all\_genes} suffix when it concerns classic DEA. 
  \item \textcolor{blue}{src/kallisto\_sleuth\_analysis.sh}:
a differential analysis is performed using the previous quantification from sleuth subsetted to readthrough and their parents to select "true" putative readthrough. Outputs is saved in \texttt{results\sleuth\_analysis} in the corresponding subdirectory with the \texttt{\_rt\_only} suffix! This script performed also intersection on what is found between DESeq2 and kallisto for the classic differential expression analysis, and generated a metagene plot as well after having called \textcolor{blue}{src/prepare\_metagene.sh} to obtain coverage at each bp from BAM.
\end{enumerate}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Differential expression analysis}

Note that the differential expression analysis was performed with DESeq2 (quantification with RSEM) and with sleuth (quantification with kallisto). Outputs from the sleuth analysis is saved in \texttt{results\sleuth\_analysis} in the corresponding subdirectory without the \texttt{\_rt\_only} suffix!
Outputs from DESeq2 analysis are saved in \texttt{results/DESeq2\_analysis} under the prefix \texttt{2017\_12\_31}. You will also find the count obtained from RSEM within the \texttt{count} directory at this place.


\end{document}
