require(tximport)
require(DESeq2)
require(ggplot2)
require(reshape2)
require(RColorBrewer)
require(pheatmap)
require(gplots)
require(data.table)

annotation <- read.table("./data/annotation.csv", sep = ";", h = T, stringsAsFactors = F) ##### annotation file 
files.names <- unname(sapply(annotation$file, function(x) strsplit(x, ".fastq", fixed = T)[[1]][1]))
gff <- fread("data/ReferenceGenomes/schizosaccharomyces_pombe.chr.gff", sep = "\t", h = F)
code <- sapply(gff$V9, function(x) paste("gene:", strsplit(strsplit(x, "gene:")[[1]][2], ";")[[1]][1],sep=""))
orig <- data.frame(chr = gff$V1, type = gff$V3, deb = gff$V4, end = gff$V5, sign = gff$V7, name = code, stringsAsFactors = F)
analysis <- list(c("cut14-208", "wt"), c("cut14-208_cdc15-118", "wt"), c("cdc15-118", "wt"), c("cut14-208_cdc15-118", "cut14-208"), c("cut14-208_cdc15-118", "cdc15-118"), c("rrp6D", "wt"), c("rrp6D", "cut14-208"))

pthresh <- 0.01
for (i in seq_along(analysis)){
  analysis.tmp <- analysis[[i]]
  save.dir <- paste(analysis.tmp[1], "_vs_", analysis.tmp[2], sep = "")
  system(paste("mkdir ./results/DESeq2_analysis/", save.dir, sep = ""))
  save.path.dir <- paste("./results/DESeq2_analysis/", save.dir, sep = "")
  ind.files <- which(annotation$strain %in% analysis.tmp)
  files.names.tmp <- files.names[ind.files]
  
  ##### Load quantification from RSEM
  files.tmp <- file.path(paste("./results/RSEM_quantif/", files.names.tmp, ".genes.results", sep = ""))

  ##### Pick count data generated by RSEM: https://bioconductor.org/packages/devel/bioc/vignettes/tximport/inst/doc/tximport.html
  names(files.tmp) <- annotation$sample[ind.files]
  txi.rsem <- tximport(files.tmp, type = "rsem")
  head(txi.rsem$counts)

  sampleTable <- data.frame(condition = annotation$strain[ind.files])
  rownames(sampleTable) <- colnames(txi.rsem$counts)

  ##### Check at log(txi.rsem$length) == 0 (i.e. log() on transcript/gene lengths equal to 0) and remove the corresponding lines: reported in https://support.bioconductor.org/p/82504/
  ind.remove <- which(apply(txi.rsem$length, 1, function(x) length(which(x == 0)))>0)
  if(length(ind.remove)>0){
    txi.rsem$counts <-txi.rsem$counts[-ind.remove, ]
    txi.rsem$length <- txi.rsem$length[-ind.remove, ]
  }   
  orig.tmp <- unlist(sapply(rownames(txi.rsem$counts), function(x) orig[which(orig$name == x), ]$type[1], simplify = F))
  
  ##### Only keep count refers to as genes 
  ind.keep.gene <- which(orig.tmp == "gene") 
  if (length(ind.keep.gene)<1) {
    print("##### Error: there is no quantification for genes! #####")
  }else {
    txi.rsem[[1]] <- txi.rsem[[1]][ind.keep.gene, ]
    txi.rsem[[2]] <- txi.rsem[[2]][ind.keep.gene, ]
    txi.rsem[[3]] <- txi.rsem[[3]][ind.keep.gene, ]
  
    ##### Input for DESeq analysis
    dds <- DESeqDataSetFromTximport(txi.rsem, sampleTable, ~condition)
  
    ##### DESeq analysis
    dds <- DESeq(dds)
    res <- results(dds) ##### Log2FC are not shrinked, see below function to get shrinked estimates
    resLFC <- lfcShrink(dds, coef=2, res=res)

    ##### Write results to txt files
    resLFC <- resLFC[order(resLFC$padj),]
    write.csv(as.data.frame(resLFC), file = paste(save.path.dir, "/deseq2_shrinked_analysis_", save.dir, ".csv", sep = ""))
  
    res.signif <- results(dds, alpha=pthresh)

    ##### VolcanoPlot 
    VolcanoPlot(res, pthresh, save.path.dir, paste("/volcano_", save.dir, ".pdf", sep = ""), save.dir) 
  
    ##### Pval hist
    PvalHist(res, pthresh,save.path.dir, paste("/padj_hist_", save.dir, ".pdf", sep = ""), save.dir)
    #Note on p-values set to NA: some values in the results table can be set to NA for one of the following reasons:
    #If within a row, all samples have zero counts, the baseMean column will be zero, and the log2 fold change estimates, p value and adjusted p value will all be set to NA.
    #If a row contains a sample with an extreme count outlier then the p value and adjusted p value will be set to NA. These outlier counts are detected by Cook’s distance. 
    #If a row is filtered by automatic independent filtering, for having a low mean normalized count, then only the adjusted p value will be set to NA. 
  
    ##### Density of upregulated genes along chromosomes
    ind.up.genes <- which(res$padj<=pthresh & res$log2FoldChange>0)
    up.genes <- rownames(res)[ind.up.genes]
    ind.orig <- unname(sapply(up.genes, function(x) which(orig$name == x & orig$type == "gene")))
    orig.up <- orig[ind.orig, ]
    up.genes.per.chr <- table(orig.up$chr)
    for (ch in names(up.genes.per.chr)){
      ind <- ind.orig[which(orig.up$chr == ch)]
      orig.tmp.up <- orig.up[ind, ]
      pos.tmp <- (orig.tmp.up$deb+orig.tmp.up$end)/2
      pos.ord.tmp <- pos.tmp[order(pos.tmp)]
      par(mar = c(5,5,2,5))
      plot(pos.tmp, res$log2FoldChange[ind.up.genes[which(orig.up$chr == ch)]], type = "h")
      par(new = T)
      hist(pos.tmp, breaks = 20, axes=F, xlab=NA, ylab=NA, cex=1.2)
      axis(side = 4)
    }
    
    ##### MA plot
    MAplotBetween2Cn(resLFC, save.path.dir, paste("/MAplot_", save.dir, ".pdf", sep = ""), save.dir) 
    #idx <- identify(resLFC$baseMean, resLFC$log2FoldChange)
    #rownames(resLFC)[idx]
    #plotCounts(dds, gene=which.min(res$padj), intgroup="condition")
    #mcols(res)$description

    ##### Heatmap of 100 Best DE genes based on pval
    rld <- rlog(dds, blind=FALSE)-
    Heatmap100DE(res, rld, pthresh, save.path.dir, paste("/Heatmap100DE_", save.dir, ".pdf", sep = ""), save.dir) 

    ##### Distance between samples 
    DistBetweenSamples(rld, save.path.dir, paste("/DistSamples_", save.dir, ".pdf", sep = "")) 
  
    ##### Dispersion plot
    plotDispEsts(dds)
  
    ##### Density of count before and after normalization
    code <- matrix(annotation$strain, nrow = 1, dimnames = list(NULL, annotation$sample))
    code.repl <- matrix(sapply(annotation$sample, function(x) strsplit(x, ".", fixed = T)[[1]][2]), nrow = 1, dimnames = list(NULL, annotation$sample))
    dds.count.norm <- counts(dds, normalize = T); dds.count <- counts(dds, normalize = F)
    count.norm <- melt(dds.count.norm , variable_name = "Samples"); count.norm$samples <- code[1, as.character(count.norm$Var2)]; count.norm$replicate <- code.repl[1, as.character(count.norm$Var2)]
    count <- melt(dds.count, variable_name = "Samples"); count$samples <- code[1, as.character(count$Var2)]; count$replicate <- code.repl[1, as.character(count$Var2)]
  
    PlotCountDen(count, save.path.dir, paste("/Raw_Count_", save.dir, ".pdf", sep = ""))
    PlotCountDen(count.norm, save.path.dir, paste("/RLENorm_Count_", save.dir, ".pdf", sep = ""))
  }
  #library("IHW")
  #resIHW <- results(dds, filterFun=ihw)
  #summary(resIHW)
}