require(tximport)
require(DESeq2)
require(ggplot2)
require(reshape2)
require(RColorBrewer)
require(pheatmap)
require(gplots)
require(data.table)
source("src/func/functions.R")

##### parameters to set by user #####
pthresh <- 0.05
pref.date.counts <- "2017_09_27"
analysis <- list(c("cut14-208", "wt"), c("cut14-208_cdc15-118", "wt"), c("cdc15-118", "wt"), c("cut14-208_cdc15-118", "cut14-208"), c("cut14-208_cdc15-118", "cdc15-118"), c("rrp6D", "wt"), c("rrp6D","cut14-208"))
ref.level <- c("wt", "wt", "wt", "cut14-208","cdc15-118", "wt", "cut14-208")


##### Code #####

##### Read annotation file
annotation <- read.table("./data/annotation.csv", sep = ";", h = T, stringsAsFactors = F) ##### annotation file 
#gtf <- fread("data/ReferenceGenomes/2017_09_19_schizosaccharomyces_pombe.chr.gtf2", sep = "\t", h = F)
gff <- fread("data/ReferenceGenomes/schizosaccharomyces_pombe.chr.gff3", sep = "\t", h = F)
gff$deb <- apply(gff, 1, function(x) if (x[7] == "+"){x[4]}else{x[5]})
gff$deb <- as.numeric(gff$deb)

##### Pick genes of interest to quantify
ind.transcripts <- which(gff$V2 == "PomBase" & gff$V3 %in% c("ncRNA_gene", "pseudogene", "gene", "snoRNA_gene", "snRNA_gene") & gff$V1 %in% c("I", "II", "III"))
gff.transcripts <- gff[ind.transcripts, ]
size <- abs(gff.transcripts$V5-gff.transcripts$V4)

##### Pick tRNA names
ind.tRNA <- which(gff$V2 == "PomBase" & gff$V3 == "tRNA_gene" & gff$V1 %in% c("I", "II", "III"))
gff.tRNA <- gff[ind.tRNA, ]

keep.names <- unname(sapply(gff.transcripts$V9, function(x) strsplit(strsplit(x, "ID=")[[1]][2], ";", fixed = T)[[1]][1]))

##### Sizes of chromosomes
sizes <- matrix( c(5579133, 4539804, 2452883), nrow = 1, dimnames = list(NULL, c("I", "II", "III")))

for (i in seq_along(analysis)){
  analysis.tmp <- analysis[[i]]
  save.dir <- paste(analysis.tmp[1], "_vs_", analysis.tmp[2], sep = "")
  system(paste("mkdir ./results/DESeq2_analysis/", save.dir, sep = ""))
  save.path.dir <- paste("./results/DESeq2_analysis/", save.dir, sep = "")
  ind.files <- which(annotation$strain %in% analysis.tmp)
  files.names.tmp <- annotation$file[ind.files]
  
  ##### Load quantification from RSEM
  files.tmp <- file.path(paste("./results/DESeq2_analysis/counts/",pref.date.counts, "_", files.names.tmp, ".genes.results", sep = ""))

  ##### Pick count data generated by RSEM: https://bioconductor.org/packages/devel/bioc/vignettes/tximport/inst/doc/tximport.html
  names(files.tmp) <- annotation$sample[ind.files]
  txi.rsem <- tximport(files.tmp, type = "rsem")

  ##### Keep genes
  ind.keep <- which(rownames(txi.rsem$counts) %in% keep.names)
  txi.rsem$length <- txi.rsem$length[ind.keep, ]
  txi.rsem$abundance <- txi.rsem$abundance[ind.keep, ]
  txi.rsem$counts <- txi.rsem$counts[ind.keep, ]
  
  sampleTable <- data.frame(condition = annotation$strain[ind.files])
  rownames(sampleTable) <- colnames(txi.rsem$counts)

  ##### Check at log(txi.rsem$length) == 0 (i.e. log() on transcript/gene lengths equal to 0) and remove the corresponding lines: reported in https://support.bioconductor.org/p/82504/
  # check to remove effective length
  ind.remove <- which(apply(txi.rsem$length, 1, function(x) length(which(x == 0)))>0)
  if(length(ind.remove)>0){
    txi.rsem$counts <- txi.rsem$counts[-ind.remove, ]
    txi.rsem$length <- txi.rsem$length[-ind.remove, ]
    txi.rsem$abundance <- txi.rsem$abundance[-ind.remove, ]
  }   

  ##### Input for DESeq analysis
  dds <- DESeqDataSetFromTximport(txi.rsem, sampleTable, ~condition)
  dds$condition <- relevel(dds$condition, ref = ref.level[i])
  
  ##### DESeq analysis
  dds <- DESeq(dds)
  res <- results(dds) ##### Log2FC are not shrinked, see below function to get shrinked estimates
  resLFC <- lfcShrink(dds, coef=2, res=res)

  ##### Write results to txt files
  resLFC <- resLFC[order(resLFC$padj),]
  write.csv(as.data.frame(resLFC), file = paste(save.path.dir, "/deseq2_shrinked_analysis_", save.dir, ".csv", sep = ""))
  
  ##### VolcanoPlot 
  VolcanoPlot(resLFC, pthresh, save.path.dir, paste("/volcano_", save.dir, ".pdf", sep = ""), save.dir) 
  
  ##### Pval hist
  PvalHist(resLFC, pthresh,save.path.dir, paste("/padj_hist_", save.dir, ".pdf", sep = ""), save.dir)
  #Note on p-values set to NA: some values in the results table can be set to NA for one of the following reasons:
  #If within a row, all samples have zero counts, the baseMean column will be zero, and the log2 fold change estimates, p value and adjusted p value will all be set to NA.
  #If a row contains a sample with an extreme count outlier then the p value and adjusted p value will be set to NA. These outlier counts are detected by Cookâ€™s distance. 
  #If a row is filtered by automatic independent filtering, for having a low mean normalized count, then only the adjusted p value will be set to NA. 
  
  # Pick data on up/downregulated genes along chromosomes
  ind.up.genes <- which(resLFC$padj<=pthresh & resLFC$log2FoldChange>0)
  up.genes <- rownames(resLFC)[ind.up.genes]
  l2fc <- sapply(up.genes, function(x) resLFC[which(rownames(resLFC) == x), ]$log2FoldChange)
  info.up.genes <- cbind(up.genes, gff.transcripts[sapply(up.genes, function(x) which(keep.names == x)), c("V1", "V4", "V5", "V7", "deb")], l2fc)
  chrom.up <- unique(info.up.genes$V1)
  
  ind.down.genes <- which(resLFC$padj<=pthresh & resLFC$log2FoldChange<0)
  down.genes <- rownames(resLFC)[ind.down.genes]
  l2fc <- sapply(down.genes, function(x) resLFC[which(rownames(resLFC) == x), ]$log2FoldChange)
  info.down.genes <- cbind(down.genes, gff.transcripts[sapply(down.genes, function(x) which(keep.names == x)), c("V1", "V4", "V5", "V7", "deb")], l2fc)
  chrom.down <- unique(info.down.genes$V1)
  
  chr.all <- unique(c(chrom.up, chrom.down))
  
  ##### Pick annotation of equally expressed genes 
  all <- setdiff(rownames(txi.rsem$counts), c(info.up.genes$up.genes, info.down.genes$up.genes))
  all <- gff.transcripts[sapply(all, function(x) which(keep.names == x)), c("V1", "V4", "V5", "V7", "deb")]
  all <- all[order(all$deb), ]
  
  ##### Implementation of tests to compare densities
  # 1/ Try binning strategy on first chromosome
  #chr <- "I"
  #bin <- round(sizes[i]/2^11)
  #w <- seq(1, sizes[i], by = bin)
  #tmp.up <- info.up.genes[which(info.up.genes$V1 == chr), ]
  #nb.to.sample <- dim(tmp.up)[1]
  #pos.genes.up <- tmp.up$deb
  #counts.up <- c()
  #for (i in 1:(length(w)-1)) {
  #  counts.up <- c(counts.up, length(which(tmp.up>= w[i] & tmp.up<w[i+1])))
  #}
  #tmp <- gff.transcripts$deb
  #counts.null <- list()
  #for (i in 1:10000) {
  #  counts <- c()
  #  tmp.test <- tmp[sample(1:dim(gff.transcripts)[1], nb.to.sample)]
  #  for (i in 1:(length(w)-1)) {
  #    counts <- c(counts, length(which(tmp.test>= w[i] & tmp.test<w[i+1])))
  #  }
  #  counts.null <- c(counts.null, list(counts))
  #}
  #counts.null.df <- do.call(rbind, counts.null)
  #pval <- sapply(seq_along(counts.up), function(x) length(which(counts.null.df[, i]>=counts.up[x]))/10000)
  
  # 2/ Compare distance to consecutive features in differentially expressed genes compared to equally expressed genes
  # Compute distance
  dist.diff <- c(); dist.all <- c()
  for (chr in chr.all) {
    tmp.up <- info.up.genes[which(info.up.genes$V1 == chr), ]$deb; 
    tmp.down <- info.down.genes[which(info.down.genes$V1 == chr), ]$deb; 
    tmp.diff <- c(tmp.up, tmp.down); tmp.diff <- tmp.diff[order(tmp.diff)]
    tmp.all <- all[which(all$V1 == chr), ]$deb; tmp.all <- tmp.all[order(tmp.all)]
    dist.diff <- c(dist.diff, tmp.diff[2:length(tmp.diff)]-tmp.diff[1:(length(tmp.diff)-1)])
    dist.all <- c(dist.all, tmp.all[2:length(tmp.all)]-tmp.all[1:(length(tmp.all)-1)])
  }
  xlim <- c(0, max(c(dist.diff, dist.all)))
  h1 <- hist(dist.diff, breaks = 100); h2 <- hist(dist.all, breaks = 100)
  ylim <- c(0, max(c(h1$counts, h2$counts)))
  
  # Fit GLM
  df.dist.2to2 <- data.frame(group = c(rep("equally.regulated", length(dist.all)), rep("up.down.regulated", length(dist.diff))), distance.2to2 = c(dist.all, dist.diff))
  glm.test.dist.2to2 <- glm(distance.2to2 ~ group, family = "poisson", data = df.dist.2to2)
  glm.test2.dist.2to2<- glm(distance.2to2 ~ group, family = "quasipoisson", data = df.dist.2to2)
  p.value.group.dist.2to2 <- coef(summary(glm.test2.dist.2to2))[,4][2]
  
  # KS test bilateral and maximum distance between Fn (repartition function)
  kstest.2to2 <- ks.test(dist.diff, dist.all)
  x0.2to2 <- ComputeDistKSTest(dist1 = dist.diff, dist2 = dist.all) 
  
  # Histograms of distance to consecutives features and Fn
  pdf(paste(save.path.dir, "/hist_dist_2to2_", save.dir, ".pdf", sep = ""), w = 10, h = 5)
  par(mfrow = c(1,2))
  hist(dist.diff, breaks = 100, main = paste("Distance between consecutive genes\nbreaks=100 - ", save.dir, sep = ""), xlab = "Distance in bp", ylim = ylim, xlim = xlim, col = adjustcolor("purple", 0.5))
  hist(dist.all, breaks = 100, col = adjustcolor("black", 0.5), add = T)
  if (kstest.2to2$p.value<0.05) {
    txt <- paste( "At a risk of 5%, distances between differentially expressed genes\nare different than equally expressed genes\n(pvalue=", format(kstest.2to2$p.value, scientific = T), ", KS bilateral)", sep = "")
  }else{
    txt <- paste( "At a risk of 5%, distances between differentially expressed genes\nare not different than equally expressed genes\n(pvalue=", format(kstest.2to2$p.value, scientific = T), ", KS bilateral)", sep = "")
  }
  if (p.value.group.dist.2to2<0.05) {
    txt <- paste(txt, "\n\nAt a risk of 5%, distances between differentially expressed genes\nare different than equally expressed genes\n(pvalue=", format(p.value.group.dist.2to2, scientific = T), ", GLM quasiPoisson)", sep = "")
  }else{
    txt <- paste(txt, "\n\nAt a risk of 5%, distances between differentially expressed genes\nare not different than equally expressed genes\n(pvalue=", format(p.value.group.dist.2to2, scientific = T), ", GLM quasiPoisson)", sep = "")
  }
  text(1.25*xlim[2]/2, ylim[2]/2, txt, cex = 0.6)
  abline(v = x0.2to2, col = "red")
  legend("topright", c("differentially expressed genes", "all but differentially expressed genes"), col = c("purple", "black"), lty = c(-1,-1), pch = c(15,15),bty = "n", cex = 0.75)
  
  plot(ecdf(dist.diff), main = "Repartition fonctions Fn", col = "purple", cex.main = 0.8, verticals = T, xlab = "Distance in bp")
  plot(ecdf(dist.all), add = T, verticals = T)
  abline(v = x0.2to2, col = "red")
  legend("bottomright", c("differentially expressed genes", "all but differentially expressed genes", "Distance where the 2 Fn are the most different (KS statistics, uni)"), col = c("purple", "black", "red"), lty = c(-1,-1, 1), pch = c(15,15,-1),bty = "n", cex = 0.6)
  dev.off()
  
  # Density along chromosomes
  pdf(paste(save.path.dir, "/density_up_down_reg_", save.dir, ".pdf", sep = ""), w = 7*length(chrom.up), h = 10)
  par(mfrow = c(2,length(chrom.up)))
  for (chr in chr.all) {
    tmp.up <- info.up.genes[which(info.up.genes$V1 == chr), ]; pos.genes.up <- unlist(apply(tmp.up, 1, function(x) x[3]:x[4]))
    tmp.down <- info.down.genes[which(info.down.genes$V1 == chr), ]; pos.genes.down <- unlist(apply(tmp.down, 1, function(x) x[3]:x[4]))
    
    den.up <- density(pos.genes.up, width = 10000); den.down <- density(pos.genes.down, width = 10000)
    pos <- c(pos.genes.up, pos.genes.down); xlim <- c(min(pos), max(pos))

    plot(den.up$x, den.up$y/max(den.up$y), type = "l", main = paste("Chromosome ", chr, " - ", save.dir, "\n window=10kb - pthresh=", pthresh, sep = ""), xlab = "Position", ylab = "Normalized density", xlim = xlim, ylim = c(-1, 1), yaxt = "n", col = "red")
    legend("topright", paste(dim(tmp.up)[1], " upregulated genes", sep = ""), lty = 1, col = "red", bty = "n")
    points(den.down$x, -den.down$y/max(den.down$y), type = "l", col = "blue")
    legend("bottomright", paste(dim(tmp.down)[1], " downregulated genes", sep = ""), lty = 1, col = "blue", bty = "n")
    abline(h=0, col = "gray")
  }
  
  for (chr in chr.all) {
    tmp.up <- info.up.genes[which(info.up.genes$V1 == chr), ]
    pos.genes.up <- unlist(apply(tmp.up, 1, function(x) x[3]:x[4]))
    tmp.down <- info.down.genes[which(info.down.genes$V1 == chr), ]
    pos.genes.down <- unlist(apply(tmp.down, 1, function(x) x[3]:x[4]))
    tmp.all <- all[which(all$V1 == chr), ]
    pos.genes.not.diff <- unlist(apply(tmp.all, 1, function(x) x[2]:x[3]))
    
    den.diff <- density(c(pos.genes.up, pos.genes.down), width = 10000);  den.not.diff <- density(pos.genes.not.diff, width = 10000)

    pos <- c(pos.genes.up, pos.genes.down, pos.genes.not.diff); xlim <- c(min(pos), max(pos))
    
    plot(den.diff$x, den.diff$y/max(den.diff$y), type = "l", main = paste("Chromosome ", chr, " - ", save.dir, "\n window=10kb - pthresh=", pthresh, sep = ""), xlab = "Position", ylab = "Normalized density", 
         xlim = xlim, ylim = c(-1, 1), yaxt = "n", col = "purple")
    legend("topright", paste(dim(tmp.up)[1]+dim(tmp.down)[1], " up/downregulated genes", sep = ""), lty = 1, col = "red", bty = "n")
    points(den.not.diff$x, -den.not.diff$y/max(den.not.diff$y), type = "l", col = "black")
    legend("bottomright", paste(dim(tmp.all)[1], " equally expressed genes", sep = ""), lty = 1, col = "red", bty = "n")
    abline(h=0, col = "gray")
  }
  dev.off()
  
  #### Distance between upregulated genes and the next tRNA
  min.dist.up.tRNA <- sapply(1:dim(info.up.genes)[1], function(x) min(abs(gff.tRNA$deb[which(gff.tRNA$V1 == info.up.genes$V1[x])]-info.up.genes$deb[x])))
  min.dist.down.tRNA <- sapply(1:dim(info.down.genes)[1], function(x) min(abs(gff.tRNA$deb[which(gff.tRNA$V1 == info.down.genes$V1[x])]-info.down.genes$deb[x])))
  min.dist.up.down.tRNA <- c(min.dist.down.tRNA, min.dist.up.tRNA)
  min.dist.equal.tRNA <- sapply(1:dim(all)[1], function(x) min(abs(gff.tRNA$deb[which(gff.tRNA$V1 == all$V1[x])]-all$deb[x])))
  
  # KS test unilateral and maximum distance between Fn 
  kstest <- ks.test(min.dist.equal.tRNA,  min.dist.up.down.tRNA, alternative = "greater")
  cdf1 <- ecdf(min.dist.equal.tRNA); cdf2 <- ecdf(min.dist.up.down.tRNA)
  x0 <- ComputeDistKSTest(dist1 = min.dist.equal.tRNA, dist2 = min.dist.up.down.tRNA) 
  
  # Fit GLM
  df.dist.tRNA <- data.frame(group = c(rep("equally.regulated", length(min.dist.equal.tRNA)), rep("up.down.regulated", length(min.dist.up.down.tRNA))), distance.tRNA = c(min.dist.equal.tRNA, min.dist.up.down.tRNA))
  glm.test.dist.tRNA <- glm(distance.tRNA ~ group, family = "poisson", data = df.dist.tRNA)
  glm.test2.dist.tRNA <- glm(distance.tRNA ~ group, family = "quasipoisson", data = df.dist.tRNA)
  p.value.group.dist.tRNA <- coef(summary(glm.test2.dist.tRNA))[,4][2]
  
  pdf(paste(save.path.dir, "/dist_next_tRNA_", save.dir, ".pdf", sep = ""), w = 10, h = 5)
  par(mfrow = c(1,2))
  ha <- hist(min.dist.equal.tRNA, breaks = 100, plot = F)
  hdiff <- hist(min.dist.up.down.tRNA, breaks = 100, plot = F)
  ylim <- c(0, max(c(hdiff$counts, ha$counts)))
  xlim <- c(0, max(c(hdiff$breaks, ha$breaks)))
  hist(min.dist.up.down.tRNA, breaks = 100, main = paste("Min distance between diff. expressed genes and tRNA promoters\nbreaks=100 - ", save.dir, sep = ""), cex.main = 0.8, xlab = "Distance in bp", ylim = ylim, xlim = xlim, col = adjustcolor("purple", 0.5))
  hist(min.dist.equal.tRNA, breaks = 100, add = T, col = adjustcolor("black", 0.25))
  legend("topright", c("differentially expressed genes", "all but differentially expressed genes"), col = c("purple", "black"), lty = c(-1,-1), pch = c(15,15),bty = "n", cex = 0.6)
  if (kstest$p.value<0.05) {
    txt <- paste("At a risk of 5%, tRNA-gene distances are lower in differentially\nexpressed genes compared to equally expressed genes\n(pvalue=", format(kstest$p.value, scientific = T), ", KS unilateral, greater)", sep = "")
  }else{
    txt <- paste("At a risk of 5%, we cannot say tRNA-gene distances are lower in\ndifferentially expressed genes compared to equally expressed genes\n(pvalue=", format(kstest$p.value, scientific = T), ", KS unilateral, greater)", sep = "")
  }
  if (p.value.group.dist.tRNA<0.05) {
    txt <- paste(txt, "\n\nAt a risk of 5%, tRNA-gene distances are lower in differentially\nexpressed genes compared to equally expressed genes\n(pvalue=", format(p.value.group.dist.tRNA, scientific = T), ", GLM quasiPoisson)", sep = "")
  }else{
    txt <- paste(txt, "\n\nAt a risk of 5%, we cannot say tRNA-gene distances are lower in\ndifferentially expressed genes compared to equally expressed genes\n(pvalue=", format(p.value.group.dist.tRNA, scientific = T), ", GLM quasiPoisson)", sep = "")
  }
  text(1.25*xlim[2]/2, ylim[2]/2, txt, cex = 0.6)
  abline(v = x0, col = "red")
  
  plot(ecdf(min.dist.up.down.tRNA), main = "Repartition fonctions Fn", col = "purple", cex.main = 0.8, verticals = T, xlab = "Distance in bp")
  plot(ecdf(min.dist.all.tRNA), add = T, verticals = T)
  abline(v = x0, col = "red")
  legend("bottomright", c("differentially expressed genes", "all but differentially expressed genes", "Distance where the 2 Fn arethe most different (KS statistics, bi)"), col = c("purple", "black", "red"), lty = c(-1,-1, 1), pch = c(15,15,-1),bty = "n", cex = 0.6)
  dev.off()

  ##### MA plot
  MAplotBetween2Cn(resLFC, save.path.dir, paste("/MAplot_", save.dir, ".pdf", sep = ""), save.dir) 
  #idx <- identify(resLFC$baseMean, resLFC$log2FoldChange)
  #rownames(resLFC)[idx]
  #plotCounts(dds, gene=which.min(res$padj), intgroup="condition")
  #mcols(res)$description

  ##### Heatmap of 100 Best DE genes based on pval
  rld <- rlog(dds, blind=FALSE)
  HeatmapDE(resLFC, rld, pthresh, save.path.dir, paste("/Heatmap100DE_", save.dir, ".pdf", sep = ""), save.dir) 

  ##### Distance between samples 
  DistBetweenSamples(rld, save.path.dir, paste("/DistSamples_", save.dir, ".pdf", sep = "")) 
  
  ##### Dispersion plot
  pdf(paste(save.path.dir, "/dispersion_estimates_", save.dir, ".pdf", sep = ""))
  plotDispEsts(dds)
  dev.off()
  
  #### Density of count before and after normalization
  code <- matrix(annotation$strain, nrow = 1, dimnames = list(NULL, annotation$sample))
  code.repl <- matrix(sapply(annotation$sample, function(x) strsplit(x, ".", fixed = T)[[1]][2]), nrow = 1, dimnames = list(NULL, annotation$sample))
  cds.count.norm <- counts(dds, normalize = T); cds.count <- counts(dds, normalize = F)
  count.norm <- melt(cds.count.norm , variable_name = "Samples"); count.norm$samples <- code[1, as.character(count.norm$Var2)]; count.norm$replicate <- code.repl[1, as.character(count.norm$Var2)]
  count <- melt(cds.count, variable_name = "Samples"); count$samples <- code[1, as.character(count$Var2)]; count$replicate <- code.repl[1, as.character(count$Var2)]
  
  PlotCountDen(count, save.path.dir, paste("/Raw_Count_", save.dir, ".pdf", sep = ""))
  PlotCountDen(count.norm, save.path.dir, paste("/Norm_Count_", save.dir, ".pdf", sep = ""))

  ##### Date results file
  system(paste("bash src/date.sh ./results/DESeq2_analysis/", save.dir, sep = ""))
  
}