require(tximport)
require(DESeq2)
require(ggplot2)
require(reshape2)
require(RColorBrewer)
require(pheatmap)
require(gplots)
require(data.table)
require(parallel)
require(MASS)
source("src/func/functions.R")

##### parameters to set by user #####
pthresh <- 0.05
pref.date.counts <- "2017_09_27"
analysis <- list(c("cut14-208", "wt"), c("cut14-208_cdc15-118", "wt"), c("cdc15-118", "wt"), c("cut14-208_cdc15-118", "cut14-208"), c("cut14-208_cdc15-118", "cdc15-118"), c("rrp6D", "wt"), c("rrp6D","cut14-208"))
ref.level <- c("wt", "wt", "wt", "cut14-208","cdc15-118", "wt", "cut14-208")
do.perm.dist.tRNA <- F 
threshLFC <- 0.5

##### Code #####

##### Read annotation file
annotation <- read.table("./data/2017_10_03_annotation.csv", sep = ";", h = T, stringsAsFactors = F) ##### annotation file 
gff <- fread("data/ReferenceGenomes/schizosaccharomyces_pombe.chr.gff3", sep = "\t", h = F)
gff$deb <- apply(gff, 1, function(x) if (x[7] == "+"){x[4]}else{x[5]})
gff$deb <- as.numeric(gff$deb)

##### Pick features of interest
ind.transcripts <- which(gff$V2 == "PomBase" & gff$V3 %in% c("ncRNA_gene", "pseudogene", "gene", "snoRNA_gene", "snRNA_gene") & gff$V1 %in% c("I", "II", "III"))
gff.transcripts <- gff[ind.transcripts, ]
write.table(gff.transcripts[, -"deb"], "data/ReferenceGenomes/2017_10_03_selected_features_s_pombe.gff3", sep = "\t", row.names = F, quote = F, col.names = F) 

##### Pick tRNA
ind.tRNA <- which(gff$V2 == "PomBase" & gff$V3 == "tRNA_gene" & gff$V1 %in% c("I", "II", "III"))
gff.tRNA <- gff[ind.tRNA, ]
write.table(gff.tRNA[, -"deb"], "data/ReferenceGenomes/2017_10_03_tRNAs_s_pombe.gff3", sep = "\t", row.names = F, col.names = F, quote = F)

##### Id of feature to keep for differential analysis
keep.names <- unname(sapply(gff.transcripts$V9, function(x) strsplit(strsplit(x, "ID=")[[1]][2], ";", fixed = T)[[1]][1]))

##### Sizes of chromosomes but MT
sizes <- matrix( c(5579133, 4539804, 2452883), nrow = 1, dimnames = list(NULL, c("I", "II", "III")))

##### Size of features
size <- abs(gff.transcripts$V5-gff.transcripts$V4)

##### Loop to perform each analysis
for (i in seq_along(analysis)){
  analysis.tmp <- analysis[[i]]
  #save.dir <- paste(analysis.tmp[1], "_vs_", analysis.tmp[2], sep = "")
  save.dir <- paste("2017_10_31_",analysis.tmp[1], "_vs_", analysis.tmp[2], sep = "")
  ##### Create one results dir per analysis
  system(paste("mkdir ./results/DESeq2_analysis/", save.dir, sep = ""))
  save.path.dir <- paste("./results/DESeq2_analysis/", save.dir, sep = "")
  ind.files <- which(annotation$strain %in% analysis.tmp)
  files.names.tmp <- annotation$file[ind.files]
  
  ##### Load quantification from RSEM
  files.tmp <- sapply(files.names.tmp, function(x) system(paste("ls ./results/DESeq2_analysis/counts/*_", x, ".genes.results", sep = ""), intern = T))
  
  ##### Pick count data generated by RSEM: https://bioconductor.org/packages/devel/bioc/vignettes/tximport/inst/doc/tximport.html
  names(files.tmp) <- annotation$sample[ind.files]
  txi.rsem <- tximport(files.tmp, type = "rsem")
  
  ##### Keep features for quantification
  ind.keep <- which(rownames(txi.rsem$counts) %in% keep.names)
  txi.rsem.keep <- c(lapply(txi.rsem[1:3], function(x) x[ind.keep, ]), list(countsFromAbundance=txi.rsem$countsFromAbundance))
  
  sampleTable <- data.frame(condition = annotation$strain[ind.files])
  rownames(sampleTable) <- colnames(txi.rsem.keep$counts)
  
  ##### Check at log(txi.rsem$length) == 0 (i.e. log() on transcript/gene lengths equal to 0) and remove the corresponding lines: reported in https://support.bioconductor.org/p/82504/
  # check to remove effective length
  ind.remove <- which(apply(txi.rsem.keep$length, 1, function(x) length(which(x == 0)))>0)
  if(length(ind.remove)>0){
    txi.rsem.keep.notnull <- c(lapply(txi.rsem.keep[1:3], function(x) x[-ind.remove, ]), list(countsFromAbundance=txi.rsem.keep$countsFromAbundance))
  }else{
    txi.rsem.keep.notnull <- txi.rsem.keep
  }   
  
  rm(txi.rsem)
  rm(txi.rsem.keep)
  
  ##### Input for DESeq analysis
  dds <- DESeqDataSetFromTximport(txi.rsem.keep.notnull, sampleTable, ~condition)
  dds$condition <- relevel(dds$condition, ref = ref.level[i])
  
  ##### DESeq analysis: are there differentially expressed genes?
  dds <- DESeq(dds, betaPrior = T)
  res <- results(dds, addMLE = T)
  resLFC <- results(dds, lfcThreshold=threshLFC, altHypothesis="greaterAbs", addMLE = T)
  
  nb.diff.res <- length(which(res$padj<pthresh))
  nb.diff.res.threshLFC <- length(which(resLFC$padj<pthresh))
    
  pdf(paste(save.path.dir, "scale_thresholdLFC_signif_", save.dir, ".pdf", sep = ""), width = 8, height = 4.5)
  par(mfrow = c(1,2))
  plotMA(res, main = paste("nb diff. expressed features\n in classic analysis is ", nb.diff.res, sep = ""), alpha = pthresh, ylim = c(-3,3))
  abline(h=0,col="dodgerblue",lwd=2)
  plotMA(resLFC, main = paste("nb diff. expressed features with LFC\nthreshold at log2(1.5)=0.58 is ", nb.diff.res.threshLFC, sep = ""), alpha = pthresh, ylim = c(-3,3))
  abline(h=c(-threshLFC, threshLFC),col="dodgerblue",lwd=2)
  dev.off()

  ##### Write results to txt files
  resLFC <- as.data.frame(resLFC[order(resLFC$padj),])
  write.table(resLFC, file = paste(save.path.dir, "/deseq2_shrinked_analysis_thresholdLFC_", threshLFC, "_", save.dir, ".txt", sep = ""), sep = "\t", col.names = T, row.names = T, quote = F)
  res <- as.data.frame(res[order(res$padj),])
  write.table(res, file = paste(save.path.dir, "/deseq2_shrinked_analysis_diff_expressed_", save.dir, ".txt", sep = ""), sep = "\t", col.names = T, row.names = T, quote = F)
 
  ##### VolcanoPlot 
  VolcanoPlot(resLFC, pthresh, save.path.dir, paste("/volcano_", save.dir, ".pdf", sep = ""), save.dir) 
  
  ##### Pval hist
  PvalHist(resLFC, pthresh,save.path.dir, paste("/padj_hist_", save.dir, ".pdf", sep = ""), save.dir)
  #Note on p-values set to NA: some values in the results table can be set to NA for one of the following reasons:
  #If within a row, all samples have zero counts, the baseMean column will be zero, and the log2 fold change estimates, p value and adjusted p value will all be set to NA.
  #If a row contains a sample with an extreme count outlier then the p value and adjusted p value will be set to NA. These outlier counts are detected by Cookâ€™s distance. 
  #If a row is filtered by automatic independent filtering, for having a low mean normalized count, then only the adjusted p value will be set to NA. 
  
  ###### Up/downregulated genes along chromosomes
  ind.up.genes <- which(resLFC$padj<pthresh & resLFC$log2FoldChange>0)
  genes <- rownames(resLFC)[ind.up.genes]
  l2fc <- sapply(genes, function(x) resLFC[which(rownames(resLFC) == x), ]$log2FoldChange)
  info.up.genes <- cbind(genes, gff.transcripts[sapply(genes, function(x) which(keep.names == x)), c("V1", "V4", "V5", "V7", "deb")], l2fc)
  chrom.up <- unique(info.up.genes$V1)
  
  ind.down.genes <- which(resLFC$padj<pthresh & resLFC$log2FoldChange<0)
  genes <- rownames(resLFC)[ind.down.genes]
  l2fc <- sapply(genes, function(x) resLFC[which(rownames(resLFC) == x), ]$log2FoldChange)
  info.down.genes <- cbind(genes, gff.transcripts[sapply(genes, function(x) which(keep.names == x)), c("V1", "V4", "V5", "V7", "deb")], l2fc)
  chrom.down <- unique(info.down.genes$V1)
  
  chr.diff.all <- unique(c(chrom.up, chrom.down))
  chr.diff.all <- chr.diff.all[order(chr.diff.all)]
  
  ##### Pick annotation of equally expressed genes 
  genes <- setdiff(rownames(txi.rsem.keep.notnull$counts), c(info.up.genes$genes, info.down.genes$genes))
  l2fc <- sapply(genes, function(x) resLFC[which(rownames(resLFC) == x), ]$log2FoldChange)
  info.equ.genes <- cbind(genes, gff.transcripts[sapply(genes, function(x) which(keep.names == x)), c("V1", "V4", "V5", "V7", "deb")],l2fc)
  
  ##### Implementation of test to compare densities/to assess presence of clusters of DE genes
  if(do.perm.dist.tRNA){
    #10kb - 25kb - 50kb
    #Calculate how many differentially expressed genes had start position within 10kb of the start position of another differentially expressed gene
    tmp.all <- rbind(info.up.genes[, c("V1", "deb")], info.down.genes[, c("V1", "deb")], info.equ.genes[, c("V1", "deb")])
    tmp.all <- tmp.all[sample(dim(tmp.all)[1]), ]
    obs <- c(); pval <- c(); N <- 10000; dist.clust.simu <- list()
    for (w in c(10000, 25000, 50000)) {
      print(w)
      dist.clust.obs <- list()
      for (chr in chr.diff.all){
        temp <- tmp.all[tmp.all$V1 == chr, ]$deb
        tmp.diff <- rbind(info.up.genes[which(info.up.genes$V1 == chr), "deb"], info.down.genes[which(info.down.genes$V1 == chr), "deb"])$deb
        nb.clust <- sapply(tmp.diff, function(x) {tmp <- setdiff(tmp.diff, x); length(which(tmp>=(x-w) & tmp <=(x+w)))})
        dist.clust.obs <- c(dist.clust.obs, list(nb.clust))
        eval(parse(text = paste("temp.chr", chr, " = temp", sep = "")))
      }
      names(dist.clust.obs) <- chr.diff.all
      obs.w <- mean(unlist(dist.clust.obs))
      obs <- c(obs, obs.w)
      
      nb.diff.chrom <- lapply(dist.clust.obs, function(x) length(x))
      cl = parallel::makeCluster(parallel::detectCores())
      clusterExport(cl, c("nb.diff.chrom","chr.diff.all","w", "N",paste("temp.chr", chr.diff.all, sep = "")))
      res = parSapply(cl, 1:N, function(x) {
        nb.clust.chr <- c()
        for (chr in chr.diff.all){
          eval(parse(text = paste("temp <- temp.chr", chr, sep = "")))
          use <- temp[sample(seq_along(temp), nb.diff.chrom[[chr]], replace = T)]
          nb.clust.chr <- c(nb.clust.chr, sapply(use, function(x) {tmp <- setdiff(use, x); length(which(tmp>=(x-w) & tmp <=(x+w)))}))
        }
        mean(nb.clust.chr)})
      dist.clust.simu <- c(dist.clust.simu,list(res))
      pval <- c(pval, length(which(res>obs.w))/N)
      stopCluster(cl)    
    }
    output_perm_density_feature <- data.frame(N = rep(N, length(pval)), w = c("10kb", "25kb", "50kb"), pval = pval, obs.clust = obs, stringsAsFactors = F)
    write.table(output_perm_density_feature, file = paste(save.path.dir, "/test_clusters_DE_features_", save.dir, ".txt", sep = ""), sep = "\t", col.names = T, row.names = F, quote = F)
  }
  
  ##### Density along chromosomes
  pdf(paste(save.path.dir, "/density_up_down_reg_", save.dir, ".pdf", sep = ""), w = 7*length(chr.diff.all), h = 10)# 12)
  par(mfrow = c(2,1))
  den.d <- NULL; pos.d <- NULL; den.nd <- NULL; pos.nd <- NULL; ud <- c(); equ <- c(); den.diffa <- c(); pos.diff <- list(); up <- c(); dw <- c()
  for (chr in chr.diff.all) {
    tmp.up <- info.up.genes[which(info.up.genes$V1 == chr), ]; pos.genes.up <- unlist(apply(tmp.up, 1, function(x) x[3]:x[4]))
    tmp.down <- info.down.genes[which(info.down.genes$V1 == chr), ]; pos.genes.down <- unlist(apply(tmp.down, 1, function(x) x[3]:x[4]))
    tmp.equ <- info.equ.genes[which(info.equ.genes$V1 == chr), ];  pos.genes.not.diff <- unlist(apply(tmp.equ, 1, function(x) x[3]:x[4]))
    den.diff <- density(c(pos.genes.up, pos.genes.down), width = 10000);  den.not.diff <- density(pos.genes.not.diff, width = 10000)
    pos <- c(pos.genes.up, pos.genes.down, pos.genes.not.diff); xlim <- c(min(pos), max(pos))
    den.d <- c(den.d, list(den.diff$y/max(den.diff$y)))
    pos.d <- c(pos.d, list(den.diff$x))
    den.nd <- c(den.nd, list(den.not.diff$y/max(den.not.diff$y)))
    pos.nd <- c(pos.nd, list(den.not.diff$x))
    ud <- c(ud, dim(tmp.up)[1]+dim(tmp.down)[1]); up <- c(up, dim(tmp.up)[1]); dw <- c(dw, dim(tmp.down)[1])
    equ <- c(equ, dim(tmp.equ)[1])
    tmp.den.diff <- density(c(pos.genes.up, pos.genes.down), from= xlim[1], to=xlim[2], width = 10000 )
    tmp.den.equ <- density(pos.genes.not.diff,  from= xlim[1], to=xlim[2], width = 10000 )
    tmp <- tmp.den.diff$y/max(tmp.den.diff$y) - tmp.den.equ$y/max(tmp.den.equ$y)
    den.diffa <- c(den.diffa, unlist(tmp))#list(tmp))
    pos.diff <- c(pos.diff, list( tmp.den.diff$x))
  }
  offset <- sizes[1, chr.diff.all]; offset <- c(0, cumsum(offset)[1:length(offset)])
  posd <- unlist(lapply(1:length(pos.d), function(x) pos.d[[x]]+offset[x]))
  posnd <- unlist(lapply(1:length(pos.nd), function(x) pos.nd[[x]]+offset[x]))
  posdiff <-  unlist(lapply(1:length(pos.diff), function(x) pos.diff[[x]]+offset[x]))
  
  plot(posd, unlist(den.d)+1, type = "l", main = paste("Density of features - Chromosomes ", paste(chr.diff.all, collapse = " "), " - ", save.dir, "\n window=10kb - pthresh=", pthresh, sep = ""), xlab = "Position", ylab = "Normalized density", 
       xlim = c(0, max(offset)), ylim = c(0, 2), yaxt = "n", col = "purple")
  axis(2, c(0,1), c(0,1), col = "black")
  axis(4, c(1,2), c(0,1), col = "purple", col.ticks = "purple", col.axis = "purple")
  
  points(posnd, unlist(den.nd), type = "l", col = 1)
  abline(h=1, col = "gray")
  u <- par("usr")
  arrows(u[1], u[3], u[1], u[4], code = 2, xpd = TRUE)
  text(offset[2:length(offset)], rep(1.9, length(offset)-1), paste(ud, " up (", up, ")\n/downregulated (", dw, ")", sep = ""), col = "darkorchid4", cex =0.8)
  text(offset[2:length(offset)], rep(0.9, length(offset)-1), paste(equ, " equally expressed", sep = ""), cex = 0.8)
  abline(v=offset, col = "darkgray", lwd = 1.5, lty = 2)
  
  plot(posdiff, den.diffa, type = "l", main = paste("Delta normalized density of diff.expressed - equ expressed features - Chromosomes ", paste(chr.diff.all, collapse = " "), " - ", save.dir, "\n window=10kb - pthresh=", pthresh, sep = ""), 
       xlab = "Position", xlim = c(0, max(offset)), ylim = c(-1, 1), col = "darkgray", ylab = "Delta normalized density")
  abline(h=0, col = "gray", lty = 2)
  abline(v=offset, col = "darkgray", lwd = 1.5, lty = 2)
  dev.off()
  
  #### Distance between upregulated genes and the next tRNA promoters
  min.dist.up.tRNA <- sapply(1:dim(info.up.genes)[1], function(x) min(abs(gff.tRNA$deb[which(gff.tRNA$V1 == info.up.genes$V1[x])]-info.up.genes$deb[x])))
  min.dist.down.tRNA <- sapply(1:dim(info.down.genes)[1], function(x) min(abs(gff.tRNA$deb[which(gff.tRNA$V1 == info.down.genes$V1[x])]-info.down.genes$deb[x])))
  min.dist.up.down.tRNA <- c(min.dist.down.tRNA, min.dist.up.tRNA)
  min.dist.equal.tRNA <- sapply(1:dim(info.equ.genes)[1], function(x) min(abs(gff.tRNA$deb[which(gff.tRNA$V1 == info.equ.genes$V1[x])]-info.equ.genes$deb[x])))
  
  # KS test unilateral and maximum distance between Fn 
  #kstest.tRNA <- ks.test(min.dist.equal.tRNA,  min.dist.up.down.tRNA, alternative = "greater")
  #x0.tRNA <- ComputeDistKSTest(dist1 = min.dist.equal.tRNA, dist2 = min.dist.up.down.tRNA) 
  
  # Fit NegBin GLM
  df.dist.tRNA <- data.frame(group = c(rep("equally.regulated", length(min.dist.equal.tRNA)), rep("up.down.regulated", length(min.dist.up.down.tRNA))), distance.tRNA = c(min.dist.equal.tRNA, min.dist.up.down.tRNA))
  nb.sample <- length(which(df.dist.tRNA$group == "up.down.regulated"))
  # subsample according to the nb of diff expressed genes
  df.tmp <- rbind(df.dist.tRNA[sample(which(df.dist.tRNA$group == "equally.regulated"), nb.sample, replace = T), ], df.dist.tRNA[which(df.dist.tRNA$group == "up.down.regulated"), ])
  #glm.nb.0.tRNA <- glm.nb(distance.tRNA ~ 1, data = df.dist.tRNA)
  glm.nb.dist.tRNA <- glm.nb(distance.tRNA ~ group, data = df.tmp)
  #comp2 <- anova(glm.nb.0.tRNA, glm.nb.dist.tRNA, test = "Chisq")["Pr(Chi)"][2,]
  pval.gr <- summary(glm.nb.dist.tRNA )[["coefficients"]][2, 4]
  
  pdf(paste(save.path.dir, "/dist_next_tRNA_", save.dir, ".pdf", sep = ""))
  ha <- hist(min.dist.equal.tRNA, breaks = 100, plot = F)
  hdiff <- hist(min.dist.up.down.tRNA, breaks = 100, plot = F)
  ylim <- c(0, max(c(hdiff$counts, ha$counts)));  xlim <- c(0, max(c(hdiff$breaks, ha$breaks)))
  
  par(mar=c(5, 4, 4, 6) + 0.1)
  hist(min.dist.up.down.tRNA, breaks = 100, main = paste("Min distance between diff. expressed genes and tRNA promoters\nbreaks=100 - ", save.dir, sep = ""), cex.main = 0.8, xlab = "Distance in bp", ylim = ylim, xlim = xlim, col = adjustcolor("purple", 0.5))
  hist(min.dist.equal.tRNA, breaks = 100, add = T, col = adjustcolor("black", 0.25))
  
  mu.null <- exp(coefficients(glm.nb.dist.tRNA)[1]); mu.1 <- exp(sum(coefficients(glm.nb.dist.tRNA)));
  xequ <- df.dist.tRNA[which(df.dist.tRNA$group == "equally.regulated"), ]$distance.tRNA; xequ <- xequ[order(xequ)]
  xdiff <- df.dist.tRNA[which(df.dist.tRNA$group == "up.down.regulated"), ]$distance.tRNA; xdiff <- xdiff[order(xdiff)]
  pred.null <- dnbinom(xequ, mu = mu.null, size = glm.nb.dist.tRNA$theta)
  pred.1 <- dnbinom(xdiff, mu = mu.1, size = glm.nb.dist.tRNA$theta)
  
  par(new = T)
  plot(xequ, pred.null, col = "black", lwd = 2, xlab = "", ylab = "", type = "l", xaxt = "n", yaxt = "n")
  points(xdiff, pred.1, col = "purple", lwd = 2, xlab = "", ylab = "", type = "l", xaxt = "n", yaxt = "n")
  axis(side = 4, ylim = c(0, 1))
  mtext(side = 4, line = 3, 'Density')
  
  legend("topright", c("differentially expressed genes", "all but differentially expressed genes"), col = c("purple", "darkgray"), lty = c(-1,-1), pch = c(15,15),bty = "n", cex = 0.6)
  if (pval.gr<0.05) {
    txt <- paste("\nAt a risk of 5%, tRNA-gene distances are different in differentially\nexpressed genes (pred=", round(mu.1), ") compared to\nequally expressed genes (pred=", round(mu.null), ")\n(pvalue, Wald test=", 
                 format(pval.gr, scientific = T), ", GLM NegBin)", sep = "")
  }else{
    txt <- paste("\nAt a risk of 5%, we cannot say tRNA-gene distances are different in\ndifferentially expressed genes (pred=", round(mu.1), ") compared to\nequally expressed genes(pred=", round(mu.null), ")\n(pvalue, Wald test=", format(pval.gr, scientific = T), 
                 ", GLM NegBin)", sep = "")
  }
  text(1.2*xlim[2]/2, max(c(pred.null, pred.1))/2, txt, cex = 0.7)
  dev.off()
  
  ##### MA plot
  MAplotBetween2Cn(resLFC, save.path.dir, paste("/MAplot_", save.dir, ".pdf", sep = ""), save.dir) 
  #idx <- identify(resLFC$baseMean, resLFC$log2FoldChange)
  #rownames(resLFC)[idx]
  #plotCounts(dds, gene=which.min(res$padj), intgroup="condition")
  #mcols(res)$description
  
  ##### Heatmap of 100 Best DE genes based on pval
  rld <- rlog(dds2, blind=FALSE)
  HeatmapDE(resLFC, rld, pthresh, save.path.dir, paste("/Heatmap100DE_", save.dir, ".pdf", sep = ""), save.dir) 
  
  ##### Distance between samples 
  DistBetweenSamples(rld, save.path.dir, paste("/DistSamples_", save.dir, ".pdf", sep = "")) 
  
  ##### Dispersion plot
  pdf(paste(save.path.dir, "/dispersion_estimates_", save.dir, ".pdf", sep = ""))
  plotDispEsts(dds2)
  dev.off()
  
  #### Density of count before and after normalization
  code <- matrix(annotation$strain, nrow = 1, dimnames = list(NULL, annotation$sample))
  code.repl <- matrix(sapply(annotation$sample, function(x) strsplit(x, ".", fixed = T)[[1]][2]), nrow = 1, dimnames = list(NULL, annotation$sample))
  cds.count.norm <- counts(dds2, normalize = T); cds.count <- counts(dds2, normalize = F)
  count.norm <- melt(cds.count.norm , variable_name = "Samples"); count.norm$samples <- code[1, as.character(count.norm$Var2)]; count.norm$replicate <- code.repl[1, as.character(count.norm$Var2)]
  count <- melt(cds.count, variable_name = "Samples"); count$samples <- code[1, as.character(count$Var2)]; count$replicate <- code.repl[1, as.character(count$Var2)]
  ##### Add chromosome informations per gene
  
  PlotCountDen(count, save.path.dir, paste("/Raw_Count_", save.dir, ".pdf", sep = ""))
  PlotCountDen(count.norm, save.path.dir, paste("/Norm_Count_", save.dir, ".pdf", sep = ""))
  
  ##### Plot log2(nom counts +1) vs condition per chromosome for differentially expressed genes
  all.diff <- rbind(info.up.genes, info.down.genes)
  count.norm.comp <- count.norm[which(count.norm$Var1 %in% all.diff$genes), ]
  count.norm.comp$chromosome <- sapply(count.norm.comp$Var1, function(x) all.diff[which(all.diff$genes == x), ]$V1)
  
  p <- ggplot(count.norm.comp, aes(factor(samples), log2(value+1))) + geom_violin() + geom_jitter(height = 0, width = 0.2) + facet_grid(~chromosome)+ylab("log2(normalized count + 1)")
  
  pdf(paste(save.path.dir, "/log2count_diff_expressed_genes_", save.dir, ".pdf", sep = ""))
  print(p)
  dev.off()
  
  ##### Date results file
  system(paste("bash src/date.sh ./results/DESeq2_analysis/", save.dir, sep = ""))
  
}
